src/type_object.rs

```diff
@@ -132,78 +136,65 @@ where
     T: PyTypeInfo,
 {
     fn type_object(py: Python) -> &PyType {
-        unsafe { py.from_borrowed_ptr(<Self as PyTypeInfo>::type_object() as *const _ as _) }
+        unsafe { py.from_borrowed_ptr(Self::type_object_raw(py) as *const _ as _) }
     }
 }
 
-/// Lazy type object for Exceptions
-#[doc(hidden)]
-pub struct LazyHeapType {
-    value: UnsafeCell<Option<NonNull<ffi::PyTypeObject>>>,
-    initialized: AtomicBool,
-}
-
-impl LazyHeapType {
-    pub const fn new() -> Self {
-        LazyHeapType {
-            value: UnsafeCell::new(None),
-            initialized: AtomicBool::new(false),
-        }
-    }
-
-    pub fn get_or_init<F>(&self, constructor: F) -> NonNull<ffi::PyTypeObject>
-    where
-        F: Fn(Python) -> NonNull<ffi::PyTypeObject>,
-    {
-        if !self
-            .initialized
-            .compare_and_swap(false, true, Ordering::Acquire)
-        {
-            // We have to get the GIL before setting the value to the global!!!
-            let gil = Python::acquire_gil();
-            unsafe {
-                *self.value.get() = Some(constructor(gil.python()));
-            }
-        }
-        unsafe { (*self.value.get()).unwrap() }
-    }
-}
-
-// This is necessary for making static `LazyHeapType`s
-//
-// Type objects are shared between threads by the Python interpreter anyway, so it is no worse
-// to allow sharing on the Rust side too.
-unsafe impl Sync for LazyHeapType {}
-
 /// Lazy type object for PyClass
 #[doc(hidden)]
 pub struct LazyStaticType {
-    value: UnsafeCell<ffi::PyTypeObject>,
-    initialized: AtomicBool,
+    // Boxed because Python expects the type object to have a stable address.
+    value: OnceCell<Box<ffi::PyTypeObject>>,
+    // Threads which have begun initialization. Used for reentrant initialization detection.
+    initializing_threads: Mutex<Option<HashSet<ThreadId>>>,
```

Comment:
Just a nit but I think `Mutex<Vec<ThreadId>>` is sufficient.

Reply:
Ah but I think `AtomicBool` is sufficient as @scalexm pointed out in the above thread because we don't release the GIL in `initialize_type_object`. Or I missed something?

Reply:
I noticed that `#[classattr]` functions can be called in `intialize_type_object`. Please ignore the above comment.

Reply:
Yes I considered `Mutex<Vec<ThreadId>>` but was a little put off by the O(N) lookup. Usually will be just one thread at most (or very) few so `Vec` should be fine 👍